Makefile

```diff
@@ -257,35 +267,46 @@ ${ISTIO_OUT}/istioctl-win.exe: depend
 	STATIC=0 GOOS=windows bin/gobuild.sh $@ istio.io/istio/pkg/version ./pilot/cmd/istioctl
 
 MIXER_GO_BINS:=${ISTIO_OUT}/mixs ${ISTIO_OUT}/mixc
-$(MIXER_GO_BINS): depend
+mixc:
+	bin/gobuild.sh ${ISTIO_OUT}/mixc istio.io/istio/pkg/version ./mixer/cmd/mixc
+mixs:
+	bin/gobuild.sh ${ISTIO_OUT}/mixs istio.io/istio/pkg/version ./mixer/cmd/mixs
+
+$(MIXER_GO_BINS):
 	bin/gobuild.sh $@ istio.io/istio/pkg/version ./mixer/cmd/$(@F)
 
-${ISTIO_OUT}/servicegraph: depend
-	bin/gobuild.sh $@ istio.io/istio/pkg/version ./mixer/example/$(@F)
+servicegraph:
+	bin/gobuild.sh $@ istio.io/istio/pkg/version ./mixer/example/servicegraph/cmd/server
+${ISTIO_OUT}/servicegraph:
+	bin/gobuild.sh $@ istio.io/istio/pkg/version ./mixer/example/servicegraph/cmd/server
 
 SECURITY_GO_BINS:=${ISTIO_OUT}/node_agent ${ISTIO_OUT}/istio_ca ${ISTIO_OUT}/multicluster_ca
-$(SECURITY_GO_BINS): depend
```

Comment:
why remove depend?

Reply:
That's the problem with using too much C-style make: even if the files are already built ( by the build step), we still call the targets in the test step. 

The 'build' target depends on 'depend' - when building individual components there is no need
to call it again and again.

Reply:
The depend was here because earlier the target was `make istio-ca` such that when run from a Mac, it would run depend as well. But with this variable (security-go-bins), I am not sure how people would go about it. May be they do make security? I am fine with that, as long as one exists, for people to build individual components.

Reply:
I see it more as a problem of using two different build systems and "go build" being a bit slow to figure out that nothing has changed.  The "depend" should just be adding the overhead of a few fstat() calls.  Overall I can't see the benefit of the changes in this section of the Makefile, and the downside is that many targets will no longer be self-sufficient.

Reply:
My undersanding: if depend is called, make believe it needs to rebuild the target. 
So 
$OUT/foo: depend 
   compile

will result in compile, even if nothing else changed.

So we ended up recompiling a lot more - even if it was already compiled in the build step.

Developers should use the simple targets ( make pilot ) - and that will recompile pilot on
every invocation.

Reply:
If I put aside the depend part aside for a moment, make isn't told about any of the actual dependencies of the go binaries.  So, there's two extremes:

1) Make can be told to just care that the file exists.  This is great for clean build performance but it's bad for devs working on a source code change because no amount of changes to source files will convince make to rebuild the artifact (the only option is to delete the file).

2) Make can be told to always rebuild.  This is essentially what needs to be done when using multiple build systems to always make sure that rebuilds reflect recent code changes (i.e., the 1st build system always needs to call into the 2nd to figure out what, if anything, needs to be rebuilt).  This is essentially what we've got today, and ideally both build systems are efficient about determining what needs to be done (which apparently isn't true for go until 1.10).

I'm not sure if the "depend" dependency is factoring in to whether make because as \#1 or \#2 (if so there's probably ways to decouple its role), but regardless it seems to me like \#2 is the most appropriate.

As for "depend", it should just depend on "init", which cares about "$(ISTIO_OUT)/istio_is_init", which just cares about "bin/init.sh" and "pilot/docker/Dockerfile.proxy_debug".  It could be that the ".PHONY" is messing up some aspect of the behavior (my very early experimentation with "order-only" dependencies didn't operate as I expected when a ".PHONY" target was specified as "order-only").  It's possible that while a 2nd "make depend" seems to correctly determine that nothing needs to be done, there could be different behaviors when something else depends on the phony "depend" (in which case "$(ISTIO_OUT)/istio_is_init" might be more appropriate than "depend"). But, I'm also doubtful that the always-rebuild behavior is a problem to be fixed (especially after golang 1.10 is released).

Reply:
I presented things as too binary on options for never-build vs. always-rebuild.  It is possible to have both schemes (and I think this is what Costin has already suggested for build systems vs devs), but there's usually fewer surprises if everyone builds the same way (and currently the release build rebuilds a few targets [mostly istioctl and the installer tars, though this might not continue in the long term] so the build might have to be careful to delete some files to make sure things are rebuilt).

Reply:
There is also a 3rd way - developers can directly use 'go build'. 

The fundamental problem is that the current makefile is slower. Even developers who type 'make pilot' and 'make docker' will end up duplicating all the go build ( which even in 1.10 needs to read
the timestamp from all the source files and read from .a and do work ). 

We can document or use some convention to make it clear which target rebuilds and which uses existing binaries - but I don't think it's ok to say 'it must always rebuild'.